include/swift/ABI/Metadata.h

```diff
@@ -120,14 +150,44 @@ struct ExternalPointer {
   StoredPointer PointerValue;
 };
 
-/// An external process's runtime target, which may be a different architecture.
+template <typename Runtime> struct WithObjCInterop {
```

Comment:
This seems a little awkward, especially with the nesting at the point of use. Would it be reasonable to make ObjC interop another parameter of `RuntimeTarget` instead?

Reply:
It's possible to sink this into a
```
template <unsigned PointerSize, bool WithObjCInterOp> struct RuntimeTarget
```

the downside is that since we need to conditionally define `TargetAnyClassMetadata`, we'd end up with 4 template specializations:
```
template <> struct RuntimeTarget<4, false>
template <> struct RuntimeTarget<4, true>
template <> struct RuntimeTarget<8, false>
template <> struct RuntimeTarget<8, true>
```

which take roughly the same amount of lines as the current definition with its reexporting `using` declarations. I like the separate nested template a little more since it doesn't duplicate any definitions and it would scale better if we have to add more dimensions, but I don't feel very strongly about  it.

Reply:
I was thinking it could be a partial specialization like `template <bool WithObjCInterOp> struct RuntimeTarget<4>`, then a little more magic to resolve `TargetAnyClassMetadata` based on the remaining template parameter. Not entirely sure if that's possible, though. If not then I agree with you that the separate nested template is better.

Reply:
I'm not a very experienced template metaprogrammer, but in my experiments it looked like template specialization and using separate nested templates (like in the patch) are the only ways to make a `using` type declaration configurable. When I discovered `std::conditional` I thought looked like a good fit, but I could not actually get it to work without instantiation failures deep in the ClassMetadata hierarchy.

Reply:
`std::conditional` ought to work, but this stuff is _really_ difficult to debug if you don't get it right.

Given all of that, let's go ahead and keep it as-is, and then if something ever comes up where we really want it to be in `RuntimeTarget` we can try again then.